api/src/auth/drivers/saml.ts

```diff
@@ -0,0 +1,171 @@
+			email,
+		};
+
+		if (userID) return userID;
+
+		// Is public registration allowed?
+		if (!allowPublicRegistration) {
+			logger.trace(`[Saml] User doesn't exist, and public registration not allowed for provider "${provider}"`);
+			throw new InvalidCredentialsException();
+		}
+
+		try {
+			await this.usersService.createOne({
+				provider,
+				first_name: userInfo.given_name,
+				last_name: userInfo.family_name,
+				email: email,
+				external_identifier: identifier,
+				role: this.config.defaultRoleId,
+				//auth_data: tokenSet.refresh_token && JSON.stringify({ refreshToken: tokenSet.refresh_token }),
+			});
+			const userID = await this.fetchUserID(identifier);
+			if (userID) return userID;
+		} catch (e) {
+			if (e instanceof RecordNotUniqueException) {
+				logger.warn(e, '[Saml] Failed to register user. User not unique');
+				throw new InvalidProviderException();
+			}
+			throw e;
+		}
+
+		throw new InvalidCredentialsException();
+	}
+
+	async login(user: User): Promise<void> {
+		return this.refresh(user);
+	}
+
+	async refresh(user: User): Promise<void> {
+		let authData = user.auth_data as AuthData;
+
+		if (typeof authData === 'string') {
+			try {
+				authData = parseJSON(authData);
+			} catch {
+				logger.warn(`[Saml] Session data isn't valid JSON: ${authData}`);
+			}
+		}
+
+		if (authData?.refreshToken) {
+			try {
+				// TODO
+				/*
+				const tokenSet = await client.refresh(authData.refreshToken);
+				// Update user refreshToken if provided
+				if (tokenSet.refresh_token) {
+					await this.usersService.updateOne(user.id, {
+						auth_data: JSON.stringify({ refreshToken: tokenSet.refresh_token }),
+					});
+				}
+*/
+			} catch (e) {
+				throw handleError(e);
+			}
+		}
+	}
+}
+
+const handleError = (e: any) => {
+	logger.trace(e, `[Saml] Unknown error`);
+	return e;
+};
+
+export function createSAMLAuthRouter(providerName: string) {
+	const router = Router();
+
+	router.get(
+		'/metadata',
+		asyncHandler(async (req, res, next) => {
+			const { sp } = getAuthProvider(providerName) as SAMLAuthDriver;
+			return res.header('Content-Type', 'text/xml').send(sp.getMetadata());
+		})
+	);
+
+	router.get(
+		'/',
+		asyncHandler(async (req, res) => {
+			const { sp, idp } = getAuthProvider(providerName) as SAMLAuthDriver;
+			const { context } = await sp.createLoginRequest(idp, 'redirect');
+			return res.redirect(context);
+		})
+	);
+
+	router.post(
+		'/acs',
+		express.urlencoded({ extended: false }),
+		asyncHandler(async (req, res, next) => {
+			const { sp, idp } = getAuthProvider(providerName) as SAMLAuthDriver;
+			const { extract } = await sp.parseLoginResponse(idp, 'post', req);
+
+			const authService = new AuthenticationService({ accountability: req.accountability, schema: req.schema });
+			const { accessToken, refreshToken, expires } = await authService.login(providerName, extract.attributes);
+
+			res.cookie(env.REFRESH_TOKEN_COOKIE_NAME, refreshToken, COOKIE_OPTIONS);
```

Comment:
Real nice moving `COOKIE_OPTIONS ` into a constant.  We should use that in the other drivers!